Cabal/Distribution/Simple/PackageIndex.hs

```diff
@@ -138,7 +143,7 @@ data PackageIndex a = PackageIndex {
   --
   -- FIXME: Clarify what "preference order" means. Check that this invariant is
   -- preserved. See #1463 for discussion.
-  packageIdIndex :: !(Map PackageName (Map Version [a]))
+  packageIdIndex :: !(Map (PackageName, Maybe UnqualComponentName) (Map Version [a]))
```

Comment:
I think that variant is OK too, but it turned out that there weren't really any cases where you wanted to find all components of a PackageName; and this version required less code changes. Notice that in all the functions below, we generally default to looking for the public library.

Actually, in your other PR (still open, but probably a little obsolete now), you did the Map as `Map PackageName (Map Version (Map (Maybe UnqualComponentName) [a])))`. I tried using *this* version first, but it's actually not the right way to go about doing things, because in a lot of ops you just want the public library...

Maybe a three level map makes sense if we decide to handle the solver more correctly than the hack I have right now, but it still doesn't make too much sense, because you still need to work out the relationships between components with the same package identifier.